lib/services/workload_identity.go

```diff
@@ -0,0 +1,120 @@
+	// CreateWorkloadIdentity creates a new WorkloadIdentity.
+	CreateWorkloadIdentity(
+		ctx context.Context, workloadIdentity *workloadidentityv1pb.WorkloadIdentity,
+	) (*workloadidentityv1pb.WorkloadIdentity, error)
+	// DeleteWorkloadIdentity deletes a SPIFFE Federation by name.
+	DeleteWorkloadIdentity(ctx context.Context, name string) error
+	// UpdateWorkloadIdentity updates a WorkloadIdentity. It will not act if the
+	// resource is not found or where the revision does not match.
+	UpdateWorkloadIdentity(
+		ctx context.Context, workloadIdentity *workloadidentityv1pb.WorkloadIdentity,
+	) (*workloadidentityv1pb.WorkloadIdentity, error)
+	// UpsertWorkloadIdentity creates or updates a WorkloadIdentity.
+	UpsertWorkloadIdentity(
+		ctx context.Context, workloadIdentity *workloadidentityv1pb.WorkloadIdentity,
+	) (*workloadidentityv1pb.WorkloadIdentity, error)
+}
+
+// MarshalWorkloadIdentity marshals the WorkloadIdentity object into a JSON byte
+// array.
+func MarshalWorkloadIdentity(
+	object *workloadidentityv1pb.WorkloadIdentity, opts ...MarshalOption,
+) ([]byte, error) {
+	return MarshalProtoResource(object, opts...)
+}
+
+// UnmarshalWorkloadIdentity unmarshals the WorkloadIdentity object from a
+// JSON byte array.
+func UnmarshalWorkloadIdentity(
+	data []byte, opts ...MarshalOption,
+) (*workloadidentityv1pb.WorkloadIdentity, error) {
+	return UnmarshalProtoResource[*workloadidentityv1pb.WorkloadIdentity](data, opts...)
+}
+
+// ValidateWorkloadIdentity validates the WorkloadIdentity object. This is
+// performed prior to writing to the backend.
+func ValidateWorkloadIdentity(s *workloadidentityv1pb.WorkloadIdentity) error {
+	switch {
+	case s == nil:
+		return trace.BadParameter("object cannot be nil")
+	case s.Version != types.V1:
+		return trace.BadParameter("version: only %q is supported", types.V1)
+	case s.Kind != types.KindWorkloadIdentity:
+		return trace.BadParameter("kind: must be %q", types.KindWorkloadIdentity)
+	case s.Metadata == nil:
+		return trace.BadParameter("metadata: is required")
+	case s.Metadata.Name == "":
+		return trace.BadParameter("metadata.name: is required")
+	case s.Spec == nil:
+		return trace.BadParameter("spec: is required")
+	case s.Spec.Spiffe.Id == "":
+		return trace.BadParameter("spec.spiffe.id: is required")
+	case !strings.HasPrefix(s.Spec.Spiffe.Id, "/"):
+		return trace.BadParameter("spec.spiffe.id: must start with a /")
+	}
+
+	for i, rule := range s.GetSpec().GetRules().GetAllow() {
+		if len(rule.Conditions) == 0 {
+			return trace.BadParameter("spec.rules.allow[%d].conditions: must be non-empty", i)
+		}
+		for j, condition := range rule.Conditions {
+			if condition.Attribute == "" {
+				return trace.BadParameter("spec.rules.allow[%d].conditions[%d].attribute: must be non-empty", i, j)
+			}
+			// Ensure exactly one operator is set.
+			operatorsSet := 0
+			if condition.Equals != "" {
+				operatorsSet++
+			}
+			if operatorsSet == 0 || operatorsSet > 1 {
```

Comment:
Is this just in preparation for more than Equals?

Reply:
Yeah - that's correct. Eventually there will be 4-5 different operators. I've just omitted them from this PR because I want to wait until the RFD is firmed up and I didn't want to balloon this PR any bigger than it already is.